core-api/src/main/java/com/optimizely/ab/config/parser/ConditionJacksonDeserializer.java

```diff
@@ -0,0 +1,129 @@
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.optimizely.ab.config.audience.AndCondition;
+import com.optimizely.ab.config.audience.AudienceIdCondition;
+import com.optimizely.ab.config.audience.Condition;
+import com.optimizely.ab.config.audience.NotCondition;
+import com.optimizely.ab.config.audience.EmptyCondition;
+import com.optimizely.ab.config.audience.NullCondition;
+import com.optimizely.ab.config.audience.OrCondition;
+import com.optimizely.ab.config.audience.UserAttribute;
+import com.optimizely.ab.internal.InvalidAudienceCondition;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class ConditionJacksonDeserializer extends JsonDeserializer<Condition> {
+    private ObjectMapper objectMapper;
+
+    public ConditionJacksonDeserializer() {
+        this(new ObjectMapper());
+    }
+
+    ConditionJacksonDeserializer(ObjectMapper objectMapper) {
+        this.objectMapper = objectMapper;
+    }
+
+    @Override
+    public Condition deserialize(JsonParser parser, DeserializationContext context) throws IOException {
+        JsonNode node = parser.getCodec().readTree(parser);
+        Condition conditions = ConditionJacksonDeserializer.<AudienceIdCondition>parseConditions(AudienceIdCondition.class, objectMapper, node);
+
+        return conditions;
+    }
+
+    private static String operand(JsonNode opNode) {
+        if (opNode != null && opNode.isTextual()) {
+            String op = opNode.asText();
+            switch (op) {
+                case "and":
+                case "or":
+                case "not":
+                    return op;
+                default:
+                    break;
+            }
+        }
+        return null;
+    }
+    protected static <T> Condition parseConditions(Class<T> clazz, ObjectMapper objectMapper, JsonNode conditionNode)
+        throws JsonProcessingException, InvalidAudienceCondition {
+
+        if (conditionNode.size() == 0) {
+            return new EmptyCondition();
+        }
+
+        List<Condition> conditions = new ArrayList<>();
+        int startingParsingIndex = 0;
+        JsonNode opNode = conditionNode.get(0);
+        String operand = operand(opNode);
+        if (operand == null) {
+            operand = "or";
+        }
+        else { // the operand is valid so move to the next node.
+            startingParsingIndex = 1;
+        }
+
+        for (int i = startingParsingIndex; i < conditionNode.size(); i++) {
+            JsonNode subNode = conditionNode.get(i);
+            if (subNode.isArray()) {
+                conditions.add(ConditionJacksonDeserializer.<T>parseConditions(clazz, objectMapper, subNode));
+            } else if (subNode.isTextual()) {
+                if (clazz != AudienceIdCondition.class) {
+                    throw new InvalidAudienceCondition(String.format("Expected AudienceIdCondition got %s", clazz.getCanonicalName()));
```

Comment:
We say "Expected AudienceIdCondition got SomeOtherType", but I think it would be more accurate to say "Strings not supported for leaves of type SomeOtherType". I.e. we trust the type that we're attempting to parse so we should complain about the datafile instead.

Same thing elsewhere.